### PR TITLE
New vehicle data - EngineOilLife feature

### PR DESCRIPTION
**Pull request by AKalinich-Luxoft**
SDL implementation of New vehicle data - EngineOilLife feature

Proposal: [New vehicle data - EngineOilLife](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0082-New-vehicle-data-EngineOilLife.md)

### CODE REVIEW

**Comment by LitvinenkoIra** [comment]
Rebuild required

**Comment by LitvinenkoIra** [comment]
Rebuild required

**Comment by LuxoftAKutsan** [comment]
@AKalinich-Luxoft @LitvinenkoIra please check build on local workstation. Fond out if there is an error in CI. Probably create docker image same on CI and investigate build.

**Comment by LitvinenkoIra** [comment]
Rebuild required

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required1

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required1

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required1

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by AKalinich-Luxoft** [comment]
Rebased due to merge conflicts

**Comment by vkushnirenko-luxoft** [comment]
test

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by vkushnirenko-luxoft** [comment]
Rebuild required

**Comment by LitvinenkoIra** [comment]
Rebased due to merge conflicts

**Comment by LitvinenkoIra** [comment]
Rebased due to merge conflicts

File: src/components/interfaces/MOBILE_API.xml [hunk 1]
```diff
@@ -4095,6 +4096,9 @@
         <param name="steeringWheelAngle" type="Boolean" mandatory="false">
             <description>Current angle of the steering wheel (in deg)</description>
         </param>
+        <param name="engineOilLife" type="Boolean" mandatory="false">
```
Comment (yang1070) [inline src/components/interfaces/MOBILE_API.xml]: parameter "engineOilLife" shall be listed under GetVehicleData/OnVehicleData/SubscribeVehicleData/UnsubscribeVehicleData rpc in group "VehicleInfo-3" in file sdl_preloaded_pt.json

Comment (AKalinich-Luxoft) [reply src/components/interfaces/MOBILE_API.xml]: @yang1070 fixed in https://github.com/smartdevicelink/sdl_core/pull/1823/commits/81f0253255e79b491f682f32cc961baacd64452e

**Comment by AKalinich-Luxoft** [comment]
Rebuild required

**Comment by theresalech** [comment]
This is a fix for #1733

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 2]
```diff
@@ -408,6 +408,44 @@ void PolicyManagerImpl::CheckPermissions(const PTString& device_id,
 #endif  // SDL_REMOTE_CONTROL
 
   cache_->CheckPermissions(groups, hmi_level, rpc, result);
+
```
Comment (jacobkeeler) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: Unrelated policy changes, please separate from this PR

File: src/components/policy/policy_regular/src/cache_manager.cc [hunk 3]
```diff
@@ -503,6 +502,13 @@ void CacheManager::CheckPermissions(const policy_table::Strings& groups,
             result.list_of_allowed_params.insert(
                 policy_table::EnumToJsonString(*params_iter));
           }
+
+          if (rpc_param.parameters.is_initialized() &&
```
Comment (jacobkeeler) [inline src/components/policy/policy_regular/src/cache_manager.cc]: Unrelated policy changes, please separate from this PR

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 4]
```diff
@@ -709,11 +709,11 @@ void PolicyManagerImpl::CheckPermissions(const PTString& app_id,
   }
 
   if (result.DisallowedInclude(rpc_params)) {
-    LOG4CXX_DEBUG(logger_, "All parameters are disallowed.");
-    result.hmi_level_permitted = kRpcUserDisallowed;
+    LOG4CXX_DEBUG(logger_, "All parameters are disallowed by user.");
+    result.hmi_level_permitted = kRpcAllParamsUserDisallowed;
```
Comment (jacobkeeler) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: Unrelated policy changes, please separate from this PR

File: src/components/application_manager/test/commands/mobile/on_way_point_change_notification_test.cc [hunk 5]
```diff
@@ -101,6 +101,15 @@ TEST_F(OnWayPointChangeNotificationTest,
...
 
+  MockAppPtr mock_app(CreateMockApp());
+
+  EXPECT_CALL(app_mngr_, application(kAppId)).WillOnce(Return(mock_app));
+
+  EXPECT_CALL(app_mngr_, CheckPolicyPermissions(_, _, _, _))
+      .WillOnce(Return(mobile_apis::Result::SUCCESS));
+
+  EXPECT_CALL(*mock_app, app_id()).WillRepeatedly(Return(kAppId));
+
```
Comment (jacobkeeler) [inline src/components/application_manager/test/commands/mobile/on_way_point_change_notification_test.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/test/commands/mobile/set_media_clock_timer_test.cc [hunk 6]
```diff
@@ -319,7 +319,9 @@ TEST_F(SetMediaClockRequestTest, OnEvent_Success) {
       hmi_apis::Common_Result::SUCCESS;
   (*msg)[am::strings::msg_params] = SmartObject(smart_objects::SmartType_Null);
 
-  EXPECT_CALL(app_mngr_, ManageMobileCommand(_, _));
+  EXPECT_CALL(
+      app_mngr_,
+      ManageMobileCommand(MobileResultCodeIs(mobile_apis::Result::SUCCESS), _));
```
Comment (jacobkeeler) [inline src/components/application_manager/test/commands/mobile/set_media_clock_timer_test.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/src/commands/mobile/subscribe_vehicle_data_request.cc [hunk 7]
```diff
@@ -34,6 +34,7 @@
 #include "application_manager/commands/mobile/subscribe_vehicle_data_request.h"
 
 #include "application_manager/application_impl.h"
+#include "application_manager/commands/command_helper.h"
```
Comment (jacobkeeler) [inline src/components/application_manager/src/commands/mobile/subscribe_vehicle_data_request.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/src/commands/mobile/get_vehicle_data_request.cc [hunk 8]
```diff
@@ -35,6 +35,7 @@
 #include "application_manager/commands/mobile/get_vehicle_data_request.h"
 
 #include "application_manager/application_impl.h"
+#include "application_manager/commands/command_helper.h"
```
Comment (jacobkeeler) [inline src/components/application_manager/src/commands/mobile/get_vehicle_data_request.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 9]
```diff
@@ -35,6 +35,7 @@
 #include "utils/macro.h"
 #include "utils/make_shared.h"
 #include "application_manager/commands/command_request_impl.h"
+#include "application_manager/commands/command_helper.h"
```
Comment (jacobkeeler) [inline src/components/application_manager/src/commands/command_request_impl.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/src/application_manager_impl.cc [hunk 10]
```diff
@@ -3349,6 +3349,7 @@ mobile_apis::Result::eType ApplicationManagerImpl::CheckPolicyPermissions(
   GetPolicyHandler().CheckPermissions(app, function_id, rpc_params, result);
 
   if (NULL != params_permissions) {
+    params_permissions->permit_result = result.hmi_level_permitted;
```
Comment (jacobkeeler) [inline src/components/application_manager/src/application_manager_impl.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/src/commands/command_helper.cc [hunk 11]
```diff
@@ -0,0 +1,201 @@
+/*
```
Comment (jacobkeeler) [inline src/components/application_manager/src/commands/command_helper.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/include/application_manager/commands/mobile/unsubscribe_vehicle_data_request.h [hunk 12]
```diff
@@ -77,6 +77,12 @@ class UnsubscribeVehicleDataRequest : public CommandRequestImpl {
    */
   bool Init() FINAL;
 
+  /**
+   * @brief Adds specific message to response info for current RPC
+   * @param response Response message, which info should be updated
+   */
+  void AddSpecificInfoToResponse(smart_objects::SmartObject& response) FINAL;
+
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/commands/mobile/unsubscribe_vehicle_data_request.h]: Unrelated changes, please separate from this PR

File: src/components/application_manager/include/application_manager/commands/mobile/subscribe_vehicle_data_request.h [hunk 13]
```diff
@@ -130,7 +136,7 @@ class SubscribeVehicleDataRequest : public CommandRequestImpl {
    * HMI
    * @param result contains result that SDL sends to mobile app.
    */
-  void CheckVISubscribtions(ApplicationSharedPtr app,
+  void CheckVISubscriptions(ApplicationSharedPtr app,
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/commands/mobile/subscribe_vehicle_data_request.h]: Unrelated changes, please separate from this PR

File: src/components/application_manager/include/application_manager/commands/mobile/get_vehicle_data_request.h [hunk 14]
```diff
@@ -64,10 +64,16 @@ class GetVehicleDataRequest : public CommandRequestImpl {
   /**
    * @brief Execute command
    **/
-  virtual void Run();
+  void Run() FINAL;
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/commands/mobile/get_vehicle_data_request.h]: Unrelated changes, please separate from this PR

File: src/components/application_manager/include/application_manager/commands/command_request_impl.h [hunk 15]
```diff
@@ -215,23 +215,10 @@ class CommandRequestImpl : public CommandImpl,
   bool CheckHMICapabilities(const mobile_apis::ButtonName::eType button) const;
 
   /**
-   * @brief Remove from current message parameters disallowed by policy table
+   * @brief Adds specific message to response info for current RPC
+   * @param response Response message, which info should be updated
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/commands/command_request_impl.h]: Unrelated changes, please separate from this PR

File: src/components/application_manager/include/application_manager/commands/command_notification_impl.h [hunk 16]
```diff
@@ -44,13 +44,46 @@ class CommandNotificationImpl : public CommandImpl {
  public:
   CommandNotificationImpl(const MessageSharedPtr& message,
                           ApplicationManager& application_manager);
+  /**
+   * @brief CommandNotificationImpl class destructor
+   **/
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/commands/command_notification_impl.h]: Unrelated changes, please separate from this PR

File: src/components/application_manager/include/application_manager/commands/command_impl.h [hunk 17]
```diff
@@ -46,6 +46,10 @@ namespace application_manager {
  * table
  */
 struct CommandParametersPermissions {
+  CommandParametersPermissions()
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/commands/command_impl.h]: Unrelated changes, please separate from this PR

File: src/components/application_manager/include/application_manager/commands/command_helper.h [hunk 18]
```diff
@@ -0,0 +1,85 @@
+/*
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/commands/command_helper.h]: Unrelated changes, please separate from this PR

File: src/components/application_manager/src/commands/command_notification_impl.cc [hunk 19]
```diff
@@ -34,10 +34,37 @@
 #include "application_manager/application_manager.h"
 #include "application_manager/message_helper.h"
 
+#include "utils/shared_ptr.h"
+#include "utils/make_shared.h"
```
Comment (jacobkeeler) [inline src/components/application_manager/src/commands/command_notification_impl.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/src/commands/mobile/unsubscribe_vehicle_data_request.cc [hunk 20]
```diff
@@ -35,6 +35,7 @@
 #include "application_manager/commands/command_impl.h"
 
 #include "application_manager/application_impl.h"
+#include "application_manager/commands/command_helper.h"
```
Comment (jacobkeeler) [inline src/components/application_manager/src/commands/mobile/unsubscribe_vehicle_data_request.cc]: Unrelated changes, please separate from this PR

File: src/components/include/application_manager/application_manager.h [hunk 21]
```diff
@@ -85,8 +85,9 @@ class EventDispatcher;
 
 class Application;
 class StateControllerImpl;
-struct CommandParametersPermissions;
```
Comment (jacobkeeler) [inline src/components/include/application_manager/application_manager.h]: Unrelated changes, please separate from this PR

File: src/components/application_manager/test/commands/mobile/on_vehicle_data_notification_test.cc [hunk 22]
```diff
@@ -130,6 +130,11 @@ TEST_F(OnVehicleDataNotificationTest,
       IviInfoUpdated(mobile_apis::VehicleDataType::VEHICLEDATA_FUELLEVEL,
                      kFuelLevel)).WillOnce(Return(applications));
 
+  EXPECT_CALL(app_mngr_, application(kAppId)).WillOnce(Return(mock_app));
```
Comment (jacobkeeler) [inline src/components/application_manager/test/commands/mobile/on_vehicle_data_notification_test.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/test/commands/mobile/on_touch_event_notification_test.cc [hunk 23]
```diff
@@ -133,6 +133,14 @@ TEST_F(OnTouchEventNotificationTest, Run_NotEmptyListOfAppsWithNavi_SUCCESS) {
   EXPECT_CALL(app_mngr_, applications_with_mobile_projection())
       .WillOnce(Return(applications_with_mobile_projection));
 
+  EXPECT_CALL(app_mngr_, application(kAppId))
```
Comment (jacobkeeler) [inline src/components/application_manager/test/commands/mobile/on_touch_event_notification_test.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/test/commands/mobile/on_tbt_client_state_notification_test.cc [hunk 24]
```diff
@@ -123,6 +123,11 @@ TEST_F(OnTBTClientStateNotificationTest,
   EXPECT_CALL(app_mngr_, applications_with_navi())
       .WillOnce(Return(applications_with_navi));
 
+  EXPECT_CALL(app_mngr_, application(kAppId)).WillOnce(Return(mock_app));
```
Comment (jacobkeeler) [inline src/components/application_manager/test/commands/mobile/on_tbt_client_state_notification_test.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/test/commands/mobile/on_hmi_status_notification_test.cc [hunk 25]
```diff
@@ -70,7 +70,10 @@ class OnHMIStatusNotificationTest
   void SetSendNotificationExpectations(MessageSharedPtr& msg) {
     EXPECT_CALL(mock_message_helper_, PrintSmartObject(_))
         .WillOnce(Return(false));
-    EXPECT_CALL(app_mngr_, SendMessageToMobile(msg, _));
```
Comment (jacobkeeler) [inline src/components/application_manager/test/commands/mobile/on_hmi_status_notification_test.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/test/commands/mobile/on_hash_change_notification_test.cc [hunk 26]
```diff
@@ -69,11 +69,15 @@ TEST_F(OnHashChangeNotificationTest, Run_ValidApp_SUCCESS) {
   std::string return_string = "1234";
   MockAppPtr mock_app = CreateMockApp();
   EXPECT_CALL(app_mngr_, application(kConnectionKey))
-      .WillOnce(Return(mock_app));
```
Comment (jacobkeeler) [inline src/components/application_manager/test/commands/mobile/on_hash_change_notification_test.cc]: Unrelated changes, please separate from this PR

File: src/components/application_manager/test/commands/command_request_impl_test.cc [hunk 27]
```diff
@@ -39,6 +39,7 @@
 #include "application_manager/commands/command_request_impl.h"
 #include "application_manager/commands/commands_test.h"
 #include "application_manager/commands/command_request_test.h"
+#include "application_manager/commands/command_helper.h"
```
Comment (jacobkeeler) [inline src/components/application_manager/test/commands/command_request_impl_test.cc]: Unrelated changes, please separate from this PR

File: src/components/interfaces/MOBILE_API.xml [hunk 28]
```diff
@@ -4112,9 +4116,7 @@
         <param name="myKey" type="Boolean" mandatory="false">
             <description>Information related to the MyKey feature</description>
         </param>
-        
         <!-- / Ford Specific Data Items -->
-
```
Comment (jacobkeeler) [inline src/components/interfaces/MOBILE_API.xml]: Unrelated style changes, unfortunately these will not be kept unless these same changes are submitted to the https://github.com/smartdevicelink/rpc_spec. The final spec will be directly copied to this Repo prior to the next release. As such, they should not be present here.

File: src/components/policy/policy_external/include/policy/policy_types.h [hunk 29]
```diff
@@ -113,7 +113,13 @@ typedef std::vector<std::string> PermissionsList;
  */
 typedef std::vector<std::string> StringArray;
 
-enum PermitResult { kRpcAllowed = 0, kRpcDisallowed, kRpcUserDisallowed };
```
Comment (jacobkeeler) [inline src/components/policy/policy_external/include/policy/policy_types.h]: Unrelated changes, please separate from this PR

File: src/components/policy/policy_external/src/cache_manager.cc [hunk 30]
```diff
@@ -1152,6 +1152,7 @@ void CacheManager::CheckPermissions(const PTString& app_id,
   policy_table::Strings::const_iterator app_groups_iter_end =
       pt_->policy_table.app_policies_section.apps[app_id].groups.end();
 
+  result.hmi_level_permitted = PermitResult::kRpcDisallowed;
```
Comment (jacobkeeler) [inline src/components/policy/policy_external/src/cache_manager.cc]: Unrelated policy changes, please separate from this PR

File: src/components/policy/policy_regular/include/policy/policy_types.h [hunk 31]
```diff
@@ -113,7 +113,13 @@ typedef std::vector<std::string> PermissionsList;
  */
 typedef std::vector<std::string> StringArray;
 
-enum PermitResult { kRpcAllowed = 0, kRpcDisallowed, kRpcUserDisallowed };
+enum PermitResult {
```
Comment (jacobkeeler) [inline src/components/policy/policy_regular/include/policy/policy_types.h]: Unrelated changes, please separate from this PR

**Comment by theresalech** [comment]
Closing in favor of #2157
